libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -355,74 +324,552 @@ exclude_symbol_in_lib(const char *name, const char *sym, const Obj_Entry *obj, c
 	return (0);
 }
 
-void *
-tramp_pgs_append(void *target, const Obj_Entry *obj, const Elf_Sym *def)
+struct tramp_pg {
+	void *_Atomic cursor;		/* Start of unused space */
+	SLIST_ENTRY(tramp_pg) link;	/* Link to next page */
+	void *trampolines[];		/* Start of trampolines */
+};
+
+static struct tramp_pg *
+tramp_pg_new(struct tramp_pg *next)
 {
-	extern const struct tramp tramp_template_exe, tramp_template_res;
-	extern const size_t sztramp_template_exe, sztramp_template_res;
+	struct tramp_pg *pg = mmap(NULL,
+	    /*
+	     * 64K is the largest size such that any capability-aligned proper
+	     * sub-range can be exactly bounded by a Morello capability.
+	     */
+	    64 * 1024,
+	    PROT_READ | PROT_WRITE | PROT_EXEC,
+	    MAP_ANON | MAP_PRIVATE,
+	    -1, 0);
+	if (pg == MAP_FAILED)
+		rtld_fatal("mmap failed");
+	atomic_store_explicit(&pg->cursor, pg->trampolines,
+	    memory_order_relaxed);
+	SLIST_NEXT(pg, link) = next;
+	return (pg);
+}
 
-	static struct tramp_pgs pgs = SLIST_HEAD_INITIALIZER(pgs);
+static void *
+tramp_pg_push(struct tramp_pg *pg, size_t len)
+{
+	void *cur = atomic_load_explicit(&pg->cursor, memory_order_relaxed);
+	void *n_cur, *tramp;
+	do {
+		tramp = cheri_setbounds(cur, len);
+		ptraddr_t n_cur_addr = roundup2(
+		    cheri_gettop(tramp), _Alignof(void *));
+		ptrdiff_t n_cur_len = cheri_gettop(cur) - n_cur_addr;
+
+		n_cur = cheri_setaddress(cur, n_cur_addr);
+		n_cur = cheri_setboundsexact(n_cur, n_cur_len);
+
+	} while (!atomic_compare_exchange_weak_explicit(
+	    &pg->cursor, &cur, n_cur,
+	    /*
+	     * Relaxed ordering is sufficient because there are no side-effects.
+	     */
+	    memory_order_relaxed, memory_order_relaxed));
+
+	assert(!cheri_gettag(tramp) || cheri_gettag(n_cur));
+	return (tramp);
+}
 
-	const struct tramp *template;
-	size_t len;
+typedef int64_t slot_index;
+
+static struct {
+	struct tramp_data *data;
+	struct tramp_map_kv {
+		_Atomic ptraddr_t key;
+		_Atomic slot_index index;
+	} *map;
+	_Atomic slot_index size, writers;
+	int exp;
+} tramp_table;
+
+static struct {
+	struct tramp_pg *_Atomic head;
+	atomic_flag lock;
+} tramp_pgs = {
+	.lock = ATOMIC_FLAG_INIT
+};
 
-	RtldLockState lockstate;
-	struct tramp_pg *pg;
-	bool did_allocate = false;
-	struct tramp *t;
+static slot_index
+tramp_table_load_factor(int exp)
+{
+	/* LOAD_FACTOR is 37.5% of capacity. */
+	return (3 << (exp - 3));
+}
 
-	if (exclude_symbol_in_lib("libc.so.7", "rfork", obj, def) ||
-	    exclude_symbol_in_lib("libc.so.7", "vfork", obj, def)) {
-		return (target);
+static struct tramp_map_kv *
+tramp_map_new(int exp)
+{
+	struct tramp_map_kv *map = xmalloc(sizeof(*map) << exp);
```

Comment:
style generally requires a blank line between variable declarations and the rest of the code.